### PR TITLE
feat(chunk estimator): Initialize chunk estimator PE-704

### PR DESCRIPTION
**Pull request by fedellen**

### CODE REVIEW

File: src/pricing/ar_data_price_chunk_estimator.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	 *
+	 * @returns Promise for a ChunkPricingInfo
+	 */
+	public async refreshPriceData(): Promise<ChunkPricingInfo> {
+		// Don't kick off another refresh while refresh is in progress
+		if (this.setupPromise) {
+			return this.setupPromise;
+		}
+
+		// Fetch the price for all values in byteVolume array and feed them into a linear regression
```
Comment (arielmelendez) [inline src/pricing/ar_data_price_chunk_estimator.ts:47]: Copy pasta

File: src/pricing/ar_data_price_chunk_estimator.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+				throw Error('Failed to generate pricing model!');
+			}
+		}
+
+		if (byteCount.valueOf() === 0) {
+			// Return base price for 0 byte uploads
+			return this.pricingInfo.baseWinstonPrice;
+		}
+
+		const numberOfChunksToUpload = Math.ceil(byteCount.valueOf() / byteCountOfChunk.valueOf());
```
Comment (arielmelendez) [inline src/pricing/ar_data_price_chunk_estimator.ts:92]: Considering adding a dividedBy function to ByteCount

File: src/pricing/ar_data_price_chunk_estimator.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+		// Lazily generate the price predictor
+		if (!this.pricingInfo) {
+			await this.refreshPriceData();
+			if (!this.pricingInfo) {
+				throw Error('Failed to generate pricing model!');
+			}
+		}
+
+		if (winston.isGreaterThan(this.pricingInfo.oneChunkWinstonPrice)) {
+			// TODO: TEST THIS UPDATED ALGO!
```
Comment (arielmelendez) [inline src/pricing/ar_data_price_chunk_estimator.ts:119]: Should this still be here?

Comment (fedellen) [reply src/pricing/ar_data_price_chunk_estimator.ts:119]: We don't use these getByteCount methods in Core/CLI at the moment, so I haven't tested that side yet. They are used in price calc though

File: src/pricing/ar_data_price_chunk_estimator.ts [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	public async getByteCountForWinston(winston: Winston): Promise<ByteCount> {
+		// Lazily generate the price predictor
+		if (!this.pricingInfo) {
+			await this.refreshPriceData();
+			if (!this.pricingInfo) {
+				throw Error('Failed to generate pricing model!');
+			}
+		}
+
+		if (winston.isGreaterThan(this.pricingInfo.oneChunkWinstonPrice)) {
```
Comment (arielmelendez) [inline src/pricing/ar_data_price_chunk_estimator.ts:118]: should be isGreaterThanOrEqualTo

File: src/pricing/ar_data_price_chunk_estimator.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+			// TODO: TEST THIS UPDATED ALGO!
+			const numberOfChunks = Math.floor(
+				+winston.minus(this.pricingInfo.baseWinstonPrice) / +this.pricingInfo.perChunkWinstonPrice
```
Comment (arielmelendez) [inline src/pricing/ar_data_price_chunk_estimator.ts:121]: Let's update the dividedBy function to take an optional rounding strategy with the default being round up and in this case here we'll round down.

File: src/pricing/ar_data_price_chunk_estimator.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+			);
+			return new ByteCount(numberOfChunks * +winston);
```
Comment (arielmelendez) [inline src/pricing/ar_data_price_chunk_estimator.ts:123]: should be times chunk size, i.e. 256KB

**Review by jdaev — APPROVED** [review]

**Review by arielmelendez — APPROVED** [review]

**Review by arielmelendez — APPROVED** [review]

**Pull request merged**